File: src/languages/amharic/amharic.ts
/**
 * Native Amharic / አማርኛ (am) text phonemizer — canonical IPA. Ethiopian Semitic, written in
 * the Ge'ez/Fidäl SYLLABARY-abugida: each codepoint is a whole CV syllable (the vowel is baked into the glyph),
 * so the g2p is a flat lookup (fidel.tsv, one Ethiopic codepoint → its CV) rather than a Brahmic matra/virama
 * engine. Two features are UNWRITTEN: GEMINATION (phonemic but unmarked — rendered single, folded vs the referee)
 * and the 6th-order vowel [ɨ], which is epenthetic and DELETED word-finally (ሁለት→hulət) and before a vowel.
 * Ejectives kʼ tʼ t͡ʃʼ pʼ t͡sʼ.
 */
import type { Phonemizer } from "../../registry.ts";
import type { SymbolData } from "../../core/normalizeSymbols.ts";
import { makeSymbolNormalizer } from "../../core/normalizeSymbols.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { makeGeezG2P } from "../../core/geez.ts";
import { MANIFEST } from "./manifest.ts";
import { makeAmharicNormalizer } from "./normalize.ts";

interface NumbersDef {
    units: string[];
    ten: string;
    teenPrefix: string;
    tens: Record<string, string>;
    hundred: string;
    thousand: string;
    million: string;
    billion: string;
}
export interface AmharicDef {
    clausePunctuation: Record<string, string>;
    numbers: NumbersDef;
    /** The shared symbol tier's data — moved verbatim, comments included. See the jsonc.
     *  Typed off `SymbolData` itself so the declaration cannot drift from what the engine reads. */
    symbolTier: Required<Pick<SymbolData, "percent" | "currency" | "units" | "magnitudes" | "exponentWords" | "multiply" | "ampersand">>;
}
const DEF = MANIFEST;
const CLAUSE_MARK = DEF.clausePunctuation;
const NUM = DEF.numbers;

/** One Amharic word → canonical IPA: fidel→CV lookup + 6th-order ɨ deletion (shared Ge'ez engine). */
export const phonemizeWord = makeGeezG2P(import.meta.url, "fidel.tsv");

// ── Numbers (decimal; Amharic) ────────────────────────────────────────────────
function numberToText(n: number): string {
    if (n < 0) return "";
    if (n < 10) return NUM.units[n]!;
    if (n === 10) return NUM.ten;
    if (n < 20) return `${NUM.teenPrefix} ${NUM.units[n - 10]}`;
    if (n < 100) {
        const t = Math.floor(n / 10), u = n % 10;
        // ⚠ THE TENS ARE KEYED "20".."90", not "2".."9". Looking up the digit alone returns undefined and the
        // ten is silently DROPPED — 25 reads "amɨst", 1998 reads thousand-nine-hundred-EIGHT.
        return NUM.tens[String(t * 10)]! + (u ? ` ${NUM.units[u]}` : "");
    }
    if (n < 1000) {
        const h = Math.floor(n / 100), r = n % 100;
        return `${h > 1 ? NUM.units[h] + " " : ""}${NUM.hundred}${r ? " " + numberToText(r) : ""}`;
    }
    if (n < 1_000_000) {
        const th = Math.floor(n / 1000), r = n % 1000;
        return `${th > 1 ? numberToText(th) + " " : ""}${NUM.thousand}${r ? " " + numberToText(r) : ""}`;
    }
    // ⚠ Scales above ሺ are European loans (ሚሊዮን / ቢሊዮን) and, unlike the bare ሺ / መቶ, KEEP their multiplier at
    // 1 — 10⁶ is አንድ ሚሊዮን. Without a composer for them the digit string is emitted raw, and the fidel g2p then
    // renders it as EMPTY IPA.
    for (const [value, scale] of [[1_000_000_000, NUM.billion], [1_000_000, NUM.million]] as const) {
        if (n >= value) {
            const q = Math.floor(n / value), r = n % value;
            return `${numberToText(q)} ${scale}${r ? " " + numberToText(r) : ""}`;
        }
    }
    return String(n);
}
function number(digits: string): string {
    const n = Number(digits);
    // ⚠ OUT OF RANGE MUST STILL BE READ. Returning `digits` leaks ASCII into the IPA — the 10¹² cap is a
    // limit of the authored magnitude words, not a reason to stop speaking. Digit-at-a-time, the same
    // fallback the fleet uses at the 2^53 cliff; see docs/investigations/bignum_fallback_investigation.md.
    if (!Number.isSafeInteger(n) || n < 0 || n >= 1e12)
        return [...digits].map((c) => numberToText(Number(c))).map(phonemizeWord).join(" ");
    return numberToText(n).split(" ").map(phonemizeWord).join(" ");
}

// Ethiopic letters (U+1200–U+135A, incl. combining marks) · Arabic digits · Ethiopic + ASCII punctuation.
//
// ⚠ THE LETTER CLASS MUST NOT REACH THE PUNCTUATION SUB-BLOCK. This is the known hazard for a script whose
// punctuation lives inside its own Unicode block — Burmese and Khmer each dropped EVERY sentence boundary that
// way. Here the class ends at ፚ = U+135A while ። ፣ ፤ ፥ ፦ ፧ ፨ are U+1362–U+1368, above the range, so the letter
// branch cannot swallow them. DO NOT widen this to the full block without moving the punctuation branch ahead
// of it.
const TOKEN = /([ሀ-ፚ]+)|(\d+)|([።፣፤፥፦፧፨.?!,;:])/gu;

export type ForeignPhonemizer = (latin: string) => string;

// በመቶ "in a hundred" is the standard percent construction, postposed; the currency and unit words are the
// standard loans, emitted in Ge'ez script and read by the ordinary fidel g2p. The per-key evidence — the
// corpus counts, the two exponent words that sit on opposite sides, why the magnitude list is load-bearing —
// travelled to amharic.jsonc with the values it explains.
const SYMBOLS = makeSymbolNormalizer({
    percent: DEF.symbolTier.percent,
    currency: DEF.symbolTier.currency,
    units: DEF.symbolTier.units,
    magnitudes: DEF.symbolTier.magnitudes,
    exponentWords: DEF.symbolTier.exponentWords,
    multiply: DEF.symbolTier.multiply,
    ampersand: DEF.symbolTier.ampersand,
});

/** Text normalization. SYMBOLS is threaded through it — the ordering is load-bearing (normalize.ts §9). */
const NORMALIZE = makeAmharicNormalizer(numberToText, SYMBOLS);

class AmharicPhonemizer implements Phonemizer {
    constructor(private foreign?: ForeignPhonemizer) {}
    text(input: string): string {
        return assembleClauses(NORMALIZE(input), TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(m[1]));
            else if (m[2]) sink.emit(number(m[2]));
            else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Amharic phonemizer. `foreign` handles embedded Latin runs. */
export function createAmharic(foreign?: ForeignPhonemizer): Phonemizer {
    return new AmharicPhonemizer(foreign);
}

File: csharp/Vernacula.Phonemizer/Languages/Amharic/Amharic.cs
/**
 * Native Amharic / አማርኛ (am) text phonemizer — canonical IPA.
 * Ported from src/languages/amharic/amharic.ts — see that file for the corpus evidence.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Amharic;

public sealed class AmharicNumbersDef
{
    public IReadOnlyList<string> Units { get; init; } = Array.Empty<string>();
    public string Ten { get; init; } = "";
    public string TeenPrefix { get; init; } = "";
    public IReadOnlyDictionary<string, string> Tens { get; init; } = new Dictionary<string, string>();
    public string Hundred { get; init; } = "";
    public string Thousand { get; init; } = "";
    public string Million { get; init; } = "";
    public string Billion { get; init; } = "";
}

public sealed class AmharicDef
{
    public IReadOnlyDictionary<string, string> ClausePunctuation { get; init; } = new Dictionary<string, string>();
    public AmharicNumbersDef Numbers { get; init; } = new();
    /** Cardinal → ordinal, for the FINAL word of a composed numeral. See the jsonc. */
    public IReadOnlyDictionary<string, string> Ordinals { get; init; } = new Dictionary<string, string>();
    /** The non-numeral, non-symbol words this layer speaks: the decimal point and the range frame. */
    public AmharicWords Words { get; init; } = new();
    /** The shared symbol tier's data — see the jsonc, where the evidence lives. */
    public AmharicSymbolTier SymbolTier { get; init; } = new();
}

public sealed class AmharicWords
{
    public string DecimalPoint { get; init; } = "";
    /** ⚠ `RangeFrom` WAS A BARE LITERAL in the replacement template on both sides — see the jsonc. */
    public string RangeFrom { get; init; } = "";
    public string RangeUntil { get; init; } = "";
}

/** Read a Latin run with another language's engine — injected from the registry. */
public delegate string ForeignPhonemizer(string latin);

public sealed class AmharicPhonemizer : ILanguage
{
    internal static readonly AmharicDef DEF = LoadManifest.Load<AmharicDef>("languages/amharic", "amharic.jsonc");
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => DEF.ClausePunctuation;
    private static AmharicNumbersDef NUM => DEF.Numbers;

    /** One Amharic word → canonical IPA: fidel→CV lookup + 6th-order ɨ deletion (shared Ge'ez engine). */
    public static readonly Func<string, string> PhonemizeWord = Geez.MakeGeezG2P("languages/amharic", "fidel.tsv");

    private static string NumberToText(double n)
    {
        if (n < 0) return "";
        if (n < 10) return NUM.Units[(int)n];
        if (n == 10) return NUM.Ten;
        if (n < 20) return $"{NUM.TeenPrefix} {NUM.Units[(int)n - 10]}";
        if (n < 100)
        {
            int t = (int)Math.Floor(n / 10), u = (int)(n % 10);
            return NUM.Tens[Js.NumberToString(t * 10)] + (u != 0 ? $" {NUM.Units[u]}" : "");
        }
        if (n < 1000)
        {
            double h = Math.Floor(n / 100), r = n % 100;
            return $"{(h > 1 ? NUM.Units[(int)h] + " " : "")}{NUM.Hundred}{(r != 0 ? " " + NumberToText(r) : "")}";
        }
        if (n < 1_000_000)
        {
            double th = Math.Floor(n / 1000), r = n % 1000;
            return $"{(th > 1 ? NumberToText(th) + " " : "")}{NUM.Thousand}{(r != 0 ? " " + NumberToText(r) : "")}";
        }
        foreach (var (value, scale) in new (double, string)[] { (1_000_000_000d, NUM.Billion), (1_000_000d, NUM.Million) })
        {
            if (n >= value)
            {
                double q = Math.Floor(n / value), r = n % value;
                return $"{NumberToText(q)} {scale}{(r != 0 ? " " + NumberToText(r) : "")}";
            }
        }
        return Js.NumberToString(n);
    }

    private static string Number(string digits)
    {
        var n = Js.Number(digits);
        if (!(double.IsInteger(n) && Math.Abs(n) <= 9007199254740991d) || n < 0 || n >= 1e12)
            return string.Join(" ", Js.CodePoints(digits).Select(c => PhonemizeWord(NumberToText(Js.Number(c)))));
        return string.Join(" ", NumberToText(n).Split(' ').Select(w => PhonemizeWord(w)));
    }

    private static readonly JsRe TOKEN = JsRegex.Compile("([ሀ-ፚ]+)|(\\d+)|([።፣፤፥፦፧፨.?!,;:])", "gu");

    private static readonly Func<string, string> SYMBOLS = NormalizeSymbols.MakeSymbolNormalizer(new SymbolData
    {
        Percent = DEF.SymbolTier.Percent,
        Currency = DEF.SymbolTier.Currency,
        Units = DEF.SymbolTier.Units,
        ExponentWords = DEF.SymbolTier.ExponentWords,
        Magnitudes = DEF.SymbolTier.Magnitudes,
        Ampersand = DEF.SymbolTier.Ampersand,
        Multiply = DEF.SymbolTier.Multiply,
    });

    /**
     * Text normalization. SYMBOLS is threaded through it — the ordering is load-bearing (normalize.ts §9).
     */
    private static readonly Func<string, string> NORMALIZE = Normalize.MakeAmharicNormalizer(NumberToText, SYMBOLS);

    private readonly ForeignPhonemizer? _foreign;

    public AmharicPhonemizer(ForeignPhonemizer? foreign = null) => _foreign = foreign;

    public string Text(string input)
    {
        return Clauses.AssembleClauses(NORMALIZE(input), TOKEN, (m, sink) =>
        {
            if (m.Groups[1].Success && m.Groups[1].Value.Length > 0) sink.Emit(PhonemizeWord(m.Groups[1].Value));
            else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0) sink.Emit(Number(m.Groups[2].Value));
            else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
            {
                if (CLAUSE_MARK.TryGetValue(m.Groups[3].Value, out var mk) && mk.Length > 0) sink.Pause(mk);
            }
        });
    }

    /** Build the Amharic phonemizer. `foreign` handles embedded Latin runs. */
    public static ILanguage CreateAmharic(ForeignPhonemizer? foreign = null) => new AmharicPhonemizer(foreign);

    internal static void RegisterSelf() =>
        Registry.Register("amharic", () => CreateAmharic(latin => Registry.ReadAsEnglish(latin)));
}

public sealed class AmharicSymbolTier
{
    public IReadOnlyList<string> Percent { get; init; } = Array.Empty<string>();
    public IReadOnlyDictionary<string, IReadOnlyList<string>> Currency { get; init; } = new Dictionary<string, IReadOnlyList<string>>();
    public IReadOnlyDictionary<string, IReadOnlyList<string>> Units { get; init; } = new Dictionary<string, IReadOnlyList<string>>();
    public ExponentWordsDef ExponentWords { get; init; } = new();
    public IReadOnlyList<string> Magnitudes { get; init; } = Array.Empty<string>();
    public string Ampersand { get; init; } = "";
    public MultiplyDef Multiply { get; init; } = null!;
}
